### DISCUSSION TITLE
Mêmes formules, plusieurs dimensions (unités)

### DISCUSSION THREAD

**Discussion by laem** [publicodes/publicodes#185] [Ideas]
Salut ! Un cas d'usage émerge chez nosgestesclimat. 

Il s'agit de manipuler plusieurs unités avec des mêmes formules. Un exemple va clarifier ça. 

```yaml

bilan = transport + logement + ...

transport = avion + voiture + ...

transport . voiture = consommation * distance * facteur d'émission

consommation = 6 l/100km

facteur d'émission = 3 kgCO2e/litre

facteur d'émission pétrole = 1,6 litrePétrole/litre

# Même chose avec transport . avion, moto, bus, etc.
```
On voit bien ici que pour compter les litres totaux de pétrole (on parle de pétrole brut ici), il faudrait répéter les mêmes opérations plus haut niveau, celles qui sont déjà écrites et applicables aux kgCO2e. C'est assez rébarbatif. Surtout que nous allons ensuite faire la même chose pour les litres de gaz (ou m3), et que le gaz peut se retrouver dans transport . voiture (GPL), mais aussi dans `transport . voiture . carburant . raffinage . [étape de raffinage]`.

> À noter, il n'y a pas d'équivalence facile entre CO2e et litres de pétrole qu'on pourrait faire au niveau de la formule finale (bilan / transport) et donc utiliser le système de conversion d'unités de publicodes, ça se fait au cas par cas assez profondément dans les formules (c'est un exemple simplifié d'un truc qu'on commence tout juste à construire). 

Je crois qu'une façon de répondre a ce besoin c'est d'introduire du [polymorphisme](https://en.wikipedia.org/wiki/Polymorphism_(computer_science)). 

C'est l'idée que la fonction somme dans un langage de programmation puisse à la fois sommer 2 et 3 (=5), mais aussi 'deux' + 'trois' = 'deuxtrois'.

L'idée c'est que transport . voiture est une fonction, qui peut s'appliquer à une empreinte climat comme à une empreinte pétrole. 

On a une colonne vertébrale de calcul commune, qu'il ne faut pas dupliquer, mais des différences locales. 

Cela dit, pas certain du fait que ça s'appelle ou que ce soit une variation du polymorphisme. J'imagine que d'autres langages gèrent déjà peut-être ça. 

- [ ] regarder ce qui se fait dans des trucs comme Wolfram Mathematica ? 

Ça fait également penser à des opérations sur des matrices, les matrices ayant dans notre cas deux dimensions, exprimées en kgCO2e et litrePétrole. Mais c'est pas évident, car on est loin d'avoir toutes les variables exprimables dans ces deux dimensions. 

Q'en pensez-vous ?

**Comment by laem** [comment]
J'itère dans ma réflexion. Mon 1er exemple manquait d'une façon concrète de récupérer le résultat du calcul. 

Pour récupérer le bilan du calcul version volume de pétrole, on pourrait appeler l'expression `bilan [litrePétrole]`. Cela rappelle alors le mécanisme composantes, qui permet d'appeler une partie nommée d'une somme `cotisation maladie [salarié]`. 

Je me demande si la fonctionnalité que je demande n'est pas juste une version plus sophistiquée d'un système d'unités. Comme le système de composantes, ça impliquerait de garder en mémoire plusieurs versions d'une variable, avec contamination des appels d'évaluation : pour calculer `bilan [pétrole]` je somme `avion [pétrole]` avec `voiture [pétrole]`, jusqu'au moment où je trouve une variable qui exprime une liste de valeurs de correspondance {[unité]: valeur}.

```yaml

bilan = transport + logement + ...

transport = avion + voiture + ...

transport . voiture = consommation * distance * facteur d'émission

consommation = 6 l/100km

facteur d'émission: 
  formules:
    - 3 kgCO2e/litre
    - 1,2 litrePétrolebrut/litreEssence

# Même chose avec transport . avion, moto, bus, etc.

```

**Comment by mquandalle** [comment]
> Ça fait également penser à des opérations sur des matrices, les matrices ayant dans notre cas deux dimensions, exprimées en kgCO2e et litrePétrole. Mais c'est pas évident, car on est loin d'avoir toutes les variables exprimables dans ces deux dimensions.

C'est ma première intuition aussi.

Actuellement le langage gère uniquement les opérations de base `+`, `×`, `>`, etc. sur des nombres, mais on pourrait faire des calculs sur d'autres types d'objets. On avait évoqué :

- Variables temporelles https://github.com/betagouv/mon-entreprise/pull/866

```js
a = [{début: "2020", fin: "2022", valeur: 10}]
b = [{début: "2020", fin: "2021", valeur: 1}, {début: "2022", fin: "2022", valeur: 2}]

c = a + b
// calculé par publicodes:
// [{début: "2020", fin: "2021", valeur: 11}, {début: "2022", fin: "2022", valeur: 12}]
```

- distribution de probabilités https://github.com/betagouv/publicodes/discussions/8#discussioncomment-728529

```js
poids d'un passager = {type: "loi normale", espérance: "80kg", variance: 10}
poids d'une valise = {type: "loi normale", espérance: "10kg", variance: 3}

poids total à transporter = poids d'un passager + poids d'une valise
// calculé par publicodes:
// {type: "loi normale", espérance: "90kg", variance: 13}
```

- chiffres significatifs https://github.com/betagouv/publicodes/issues/148#issuecomment-1023069389

```js
hauteur = {valeur: 1.2313, précision: 4}
largeur = {valeur: 2.3, précision: 1}

surface = hauteur × largeur
// calculé par publicodes:
// {valeur: 2.8, précision: 1}
```


Dans ton exemple les objets manipulés aurait un attribut `pétrole` et un attribut `co2eq`. Le fait que toutes les variables ne soient pas exprimées dans ces deux dimensions n'est pas vraiment gênant, on peut additionner/multiplier un scalaire par une matrice pour retomber sur nos pattes :

```js
facteur émission = { co2eq: "3 kgCO2e/litre", pétrole: "1,2 litrePétrolebrut/litre" }
consommation = "6l/100km"
distance = "300km"

impact = consommation * distance * facteur émission
// calculé par publicodes
// { co2eq: "54 kgCO2e", pétrole: "21,6 litrePétrolebrut/litre" }
```

L'implémentation pour les variables temporelles ressemblait fortement à ça, mais elle compliquait vraiment la lecture du code des mécanismes (et détériorait les performances à l'évaluation) et on avait conclu qu'il valait mieux d'abord simplifier le code du moteur avant d'ajouter cette fonctionnalité https://github.com/betagouv/mon-entreprise/pull/1579.

Il y a peu être un moyen plus “hacky” d'implémenter cette fonctionnalité sans en passer par cette abstraction d'opérations de base sur des vecteurs.

**Reply by mquandalle** [reply]
Pour l'approche plus “hacky”, peut-être lancer 2 fois le calcul avec un remplacement ?
